### PR TITLE
errors: add error types and code; modify context and some appgw errors

### PR DESCRIPTION
**Pull request by akshaysngupta**
This PR adds a new error type with ErrorCode type.
Some on this has been prototyped in Context and appgw package.

The error code has then been used for selected errors to emitted through prometheus.
Example prometheus metric emitted:
```
# HELP appgw_ingress_controller_error_counter This guage changes represents an error on AGIC
# TYPE appgw_ingress_controller_error_counter counter
appgw_ingress_controller_error_counter{controller_appgw_name="applicationgateway7ae9",controller_appgw_resource_group="resgp",controller_appgw_subscription="xxxx",controller_class="azure/application-gateway",controller_namespace="",controller_pod="",controller_version="1.1.0/f6ff0369/2020-03-25-21:07T+0000",error_code="ErrorFetchingEnpdoints"} 3
```

### CODE REVIEW

File: pkg/k8scontext/context.go [hunk 1]
```diff
@@ -449,9 +495,13 @@ func (c *Context) UpdateIngressStatus(ingressToUpdate v1beta1.Ingress, newIP IPA
 	ingress.Status.LoadBalancer.Ingress = loadBalancerIngresses
 
 	if _, err := ingressClient.UpdateStatus(ingress); err != nil {
-		errorLine := fmt.Sprintf("Unable to update ingress %s/%s status: error %s", ingress.Namespace, ingress.Name, err)
-		glog.Error(errorLine)
-		return ErrorUnableToUpdateIngress
+		e := controllererrors.NewErrorWithInnerErrorf(
+			controllererrors.ErrorUpdatingIngressStatus,
+			err,
+			"Unable to update ingress %s/%s status:", ingress.Namespace, ingress.Name,
```
Comment (3quanfeng) [inline pkg/k8scontext/context.go]: "Unable to update ingress %s/%s status"

File: pkg/metricstore/metricstore.go [hunk 2]
```diff
@@ -137,6 +156,11 @@ func (ms *AGICMetricStore) IncArmAPICallCounter() {
 	ms.armAPICallCounter.Inc()
 }
 
+// IncErrorCount increases the counter for success on ARM
```
Comment (3quanfeng) [inline pkg/metricstore/metricstore.go]: counter for error?

File: pkg/appgw/validators.go [hunk 3]
```diff
@@ -85,12 +68,20 @@ func validateURLPathMaps(eventRecorder record.EventRecorder, config *n.Applicati
...
+				)
+				return e
 			}
 
 			if validRedirect && validBackend || !validRedirect && !validBackend {
 				// TODO(draychev): Emit an event for the appropriate object
-				return validationErrors[errKeyEitherDefaults]
+				e := controllererrors.NewError(
+					controllererrors.ErrorEitherDefaults,
+					"message",
```
Comment (3quanfeng) [inline pkg/appgw/validators.go]: Maybe need a separate constructor for NewError without message?

File: pkg/k8scontext/context.go [hunk 4]
```diff
@@ -130,7 +131,12 @@ func (c *Context) Run(stopChannel chan struct{}, omitCRDs bool, envVariables env
 	var hasSynced []cache.InformerSynced
 
 	if c.informers == nil {
-		return ErrorInformersNotInitialized
+		e := controllererrors.NewError(
+			controllererrors.ErrorInformersNotInitialized,
+			"informers are not initialized",
```
Comment (3quanfeng) [inline pkg/k8scontext/context.go]: The "code" here is kind of duplicate with what your "message" tries to tell, we might just keep code for NewError()?

File: pkg/k8scontext/context.go [hunk 5]
```diff
@@ -167,7 +173,12 @@ func (c *Context) Run(stopChannel chan struct{}, omitCRDs bool, envVariables env
 
 	glog.V(1).Infoln("Waiting for initial cache sync")
 	if !cache.WaitForCacheSync(stopChannel, hasSynced...) {
-		return ErrorFailedInitialCacheSync
+		e := controllererrors.NewError(
```
Comment (3quanfeng) [inline pkg/k8scontext/context.go]: Give how the NewError is being used, the message is kind of duplicated with the code, the code is also a string underneath? we could just keep the Code field for NewError

Comment (akshaysngupta) [reply pkg/k8scontext/context.go]: for now, I have added human readable messages for each of the errors.

**Pull request merged**